### PR TITLE
Downloader functional tests

### PR DESCRIPTION
**Pull request by pnowakowski**
This change is [Reviewable](https://reviewable.io/reviews/heremaps/oss-review-toolkit/98)

### CODE REVIEW

**Comment by mnonnenmacher** [comment]
r1:
Change title to "downloader: Use only permitted chars for directory names".
Also the explanation could be improved, e.g.:
"Package names are used to create directories, which fails when they contain characters which are not allowed for directory names (e.g. "project :module_name").
Add a dynamic property normalizedName to the Package class that contains only characters which are valid for directory names, ..."

r2:
Change title to "Resolve #92: Add functional tests for downloader"
See: https://help.github.com/articles/closing-issues-using-keywords/

Also the tag is not new anymore.

---

Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[downloader/src/funTest/kotlin/DirectoryTest.kt, line 40 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98#-L-ajDE8rSrJGz3e7-t-:-L-ajDE8rSrJGz3e7-t0:b-lj6tdp) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/ecafbbc774b3044d1f1c39f6910ebf7df87d476d/downloader/src/funTest/kotlin/DirectoryTest.kt#L40)):*
> gradle
Gradle

---

*[downloader/src/funTest/kotlin/DownloaderTest.kt, line 59 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98#-L-ajr52ycLnyZGrhh8D:-L-ajr52ycLnyZGrhh8E:b-m18kwo) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/ecafbbc774b3044d1f1c39f6910ebf7df87d476d/downloader/src/funTest/kotlin/DownloaderTest.kt#L59)):*
> ```Kotlin
>                     vcsProvider = "Git",
>                     vcsUrl = "https://github.com/heremaps/oss-review-toolkit.git",
>                     vcsRevision = "",
> ```

Please add a revision here.

---

*[downloader/src/funTest/kotlin/DownloaderTest.kt, line 69 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98#-L-ajuFTekGak8c-xUbL:-L-ajuFTekGak8c-xUbM:b-2358m) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/ecafbbc774b3044d1f1c39f6910ebf7df87d476d/downloader/src/funTest/kotlin/DownloaderTest.kt#L69)):*
> ```Kotlin
>             workingDirectory.isValid() shouldBe true
>             outputDirList.indexOf(submoduleName) should beGreaterThan(-1)
>             outputDirList.indexOf("downloader") shouldBe -1
> ```

Please also check, that the created directory is a Git repository, and that the correct revision was cloned. Same for the tests below.

---

*[downloader/src/funTest/kotlin/DownloaderTest.kt, line 97 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98#-L-akIvWWmihsQBRxBZD:-L-akIvXcs7NiSzRM1oM:bmr8xq) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/ecafbbc774b3044d1f1c39f6910ebf7df87d476d/downloader/src/funTest/kotlin/DownloaderTest.kt#L97)):*
> ```Kotlin
>         }.config(tags = setOf(Expensive))
> 
>         "Downloads single revision of package from git" {
> ```

This test can be merged with the one above if we check for the correct revision there.

---

*[downloader/src/funTest/kotlin/DownloaderTest.kt, line 111 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98#-L-akXvBj5C_rGtOUzLj:-L-akXvBj5C_rGtOUzLk:bueyqtd) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/ecafbbc774b3044d1f1c39f6910ebf7df87d476d/downloader/src/funTest/kotlin/DownloaderTest.kt#L111)):*
> ```Kotlin
>                     vcsProvider = "Git",
>                     vcsUrl = "https://github.com/heremaps/oss-review-toolkit.git",
>                     vcsRevision = "",
> ```

"revision" should be used here.

---

*[downloader/src/funTest/kotlin/DownloaderTest.kt, line 136 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98#-L-akbH7hh6Yc37Hi7cX:-L-akbH7hh6Yc37Hi7cY:bueyqtd) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/ecafbbc774b3044d1f1c39f6910ebf7df87d476d/downloader/src/funTest/kotlin/DownloaderTest.kt#L136)):*
> ```Kotlin
>                     vcsProvider = "Mercurial",
>                     vcsUrl = "https://bitbucket.org/creaceed/mercurial-xcode-plugin",
>                     vcsRevision = "",
> ```

"revision" should be used here.

---

*[downloader/src/funTest/kotlin/DownloaderTest.kt, line 150 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98#-L-akgI7vS93_egsJWDC:-L-akgI7vS93_egsJWDD:b-a0649t) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/ecafbbc774b3044d1f1c39f6910ebf7df87d476d/downloader/src/funTest/kotlin/DownloaderTest.kt#L150)):*
> ```Kotlin
>             = vcs.getWorkingDirectory(File(outputDir, "${pkg.normalizedName}/${pkg.version}"))
> 
> }
> ```

I would like to separate the tests for each VCS in a separate test class to improve overview, and to make sure that the same features are tested for each VCS. Check the proposal below, should be easy to adapt:

GitTest {
"Git should clone HEAD if no revision is provided"
"Git should clone correct revision" - with revision=sha1
"Git should clone correct branch" - with revision=some branch name
"Git should clone correct tag" - with revision=some tag name

And a test for the fallback:

"Git should guess correct tag" - when the provided revision is invalid but a tag can be guessed from the version
}

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98#-:-L-ahC2f1GHQ1nukymPo:b-zajf7s)*

**Comment by mnonnenmacher** [comment]
Review status: 0 of 5 files reviewed at latest revision, 8 unresolved discussions.

---

*[model/src/main/kotlin/Package.kt, line 141 at r1](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98#-L-amXm7pIlyTkJIkzui:-L-amXm7pIlyTkJIkzuj:bqujcpp) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/c38c4cda274e186319ee1a9e3e3ca4614f9f27b2/model/src/main/kotlin/Package.kt#L141)):*
> ```Kotlin
> 
>     /**
>      * The normalized package name, so it can be used as downloaded package directory.
> ```

The normalized package name.

@see normalizePackageName

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98)*

**Comment by mnonnenmacher** [comment]
Review status: 0 of 5 files reviewed at latest revision, 8 unresolved discussions.

---

*[model/src/main/kotlin/Package.kt, line 141 at r1](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98#-L-amXm7pIlyTkJIkzui:-L-an4auhQgS1hjLVnP3:b-remqpj) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/c38c4cda274e186319ee1a9e3e3ca4614f9f27b2/model/src/main/kotlin/Package.kt#L141)):*
<details><summary><i>Previously, mnonnenmacher (Martin Nonnenmacher) wrote…</i></summary><blockquote>

The normalized package name.

@see normalizePackageName
</blockquote></details>

Overlooked that you already have the @see :)
Would maybe rephrase to "The normalized package name, can be used to create directories."

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98)*

**Comment by mnonnenmacher** [comment]
Reviewed 3 of 3 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98)*

**Comment by pnowakowski** [comment]
Done.

---

Review status: 2 of 5 files reviewed at latest revision, 8 unresolved discussions.

---

*[downloader/src/funTest/kotlin/DirectoryTest.kt, line 40 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98#-L-ajDE8rSrJGz3e7-t-:-L-bA-euUezrRi9tCPJZ:b-896fix) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/ecafbbc774b3044d1f1c39f6910ebf7df87d476d/downloader/src/funTest/kotlin/DirectoryTest.kt#L40)):*
<details><summary><i>Previously, mnonnenmacher (Martin Nonnenmacher) wrote…</i></summary><blockquote>

> gradle
Gradle
</blockquote></details>

Done.

---

*[downloader/src/funTest/kotlin/DownloaderTest.kt, line 59 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98#-L-ajr52ycLnyZGrhh8D:-L-bACuPU9x3tlys70qc:b-896fix) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/ecafbbc774b3044d1f1c39f6910ebf7df87d476d/downloader/src/funTest/kotlin/DownloaderTest.kt#L59)):*
<details><summary><i>Previously, mnonnenmacher (Martin Nonnenmacher) wrote…</i></summary><blockquote>

Please add a revision here.
</blockquote></details>

Done.

---

*[downloader/src/funTest/kotlin/DownloaderTest.kt, line 69 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98#-L-ajuFTekGak8c-xUbL:-L-bATdq1FzQqNuWjOoc:b-mupkt9) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/ecafbbc774b3044d1f1c39f6910ebf7df87d476d/downloader/src/funTest/kotlin/DownloaderTest.kt#L69)):*
<details><summary><i>Previously, mnonnenmacher (Martin Nonnenmacher) wrote…</i></summary><blockquote>

Please also check, that the created directory is a Git repository, and that the correct revision was cloned. Same for the tests below.
</blockquote></details>

`workingDirectory.isValid() shouldBe true` checks if directory is git repository, doesn't it?
Added check of revision.

---

*[downloader/src/funTest/kotlin/DownloaderTest.kt, line 97 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98#-L-akIvWWmihsQBRxBZD:-L-bBA85rhPNpxe95YXv:b-896fix) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/ecafbbc774b3044d1f1c39f6910ebf7df87d476d/downloader/src/funTest/kotlin/DownloaderTest.kt#L97)):*
<details><summary><i>Previously, mnonnenmacher (Martin Nonnenmacher) wrote…</i></summary><blockquote>

This test can be merged with the one above if we check for the correct revision there.
</blockquote></details>

Done.

---

*[downloader/src/funTest/kotlin/DownloaderTest.kt, line 111 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98#-L-akXvBj5C_rGtOUzLj:-L-bBBY4gc122G97mmRL:bod5jvt) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/ecafbbc774b3044d1f1c39f6910ebf7df87d476d/downloader/src/funTest/kotlin/DownloaderTest.kt#L111)):*
<details><summary><i>Previously, mnonnenmacher (Martin Nonnenmacher) wrote…</i></summary><blockquote>

"revision" should be used here.
</blockquote></details>

Done. (Actually now it's used in testcase where sparse checkout is being checked, called now: "Downloads subpath of package at specified revision from git")

---

*[downloader/src/funTest/kotlin/DownloaderTest.kt, line 136 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98#-L-akbH7hh6Yc37Hi7cX:-L-bBzhR2insl_LuaBRs:b-nyd9ay) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/ecafbbc774b3044d1f1c39f6910ebf7df87d476d/downloader/src/funTest/kotlin/DownloaderTest.kt#L136)):*
<details><summary><i>Previously, mnonnenmacher (Martin Nonnenmacher) wrote…</i></summary><blockquote>

"revision" should be used here.
</blockquote></details>

In this case revision should be determined for given version before projects is checked out from mercurial. 

---

*[downloader/src/funTest/kotlin/DownloaderTest.kt, line 150 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98#-L-akgI7vS93_egsJWDC:-L-bDdM1fxOOY-XZaTN8:bxexa9u) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/ecafbbc774b3044d1f1c39f6910ebf7df87d476d/downloader/src/funTest/kotlin/DownloaderTest.kt#L150)):*
<details><summary><i>Previously, mnonnenmacher (Martin Nonnenmacher) wrote…</i></summary><blockquote>

I would like to separate the tests for each VCS in a separate test class to improve overview, and to make sure that the same features are tested for each VCS. Check the proposal below, should be easy to adapt:

GitTest {
"Git should clone HEAD if no revision is provided"
"Git should clone correct revision" - with revision=sha1
"Git should clone correct branch" - with revision=some branch name
"Git should clone correct tag" - with revision=some tag name

And a test for the fallback:

"Git should guess correct tag" - when the provided revision is invalid but a tag can be guessed from the version
}
</blockquote></details>

I wasn't very happy  with how the tests are grouped, but didn't have a better idea. 
We already have MercurialTest, but there, Mercurial is being called directly, instead of with`Main.download(pkg, outputDir)` like it is here. 
I think we should either call Mercurial via `Main` class,  or call Git directly in here as well. 
Then we should add some tests for `Main.download(pkg, outputDir)` to cover cases that are not covered by Git and Mercurial tests.
I like refactoring MercurialTest option better. What do you think? 

---

*[model/src/main/kotlin/Package.kt, line 141 at r1](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98#-L-amXm7pIlyTkJIkzui:-L-bG_wx6tTKMlYd1a9a:bs7wumb) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/c38c4cda274e186319ee1a9e3e3ca4614f9f27b2/model/src/main/kotlin/Package.kt#L141)):*
<details><summary><i>Previously, mnonnenmacher (Martin Nonnenmacher) wrote…</i></summary><blockquote>

Overlooked that you already have the @see :)
Would maybe rephrase to "The normalized package name, can be used to create directories."
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98#-:-L-b9V5YM9nLiH01kGcw:b-896fix)*

**Comment by mnonnenmacher** [comment]
Can you please confirm that the expensive tests succeed, because they are not run in Travis CI? I would approve this PR then.

---

Reviewed 1 of 3 files at r3, 1 of 2 files at r4.
Review status: 4 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[downloader/src/funTest/kotlin/DownloaderTest.kt, line 69 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98#-L-ajuFTekGak8c-xUbL:-L-bIU10IHWfwKZj94io:b-80bnvp) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/ecafbbc774b3044d1f1c39f6910ebf7df87d476d/downloader/src/funTest/kotlin/DownloaderTest.kt#L69)):*
<details><summary><i>Previously, pnowakowski wrote…</i></summary><blockquote>

`workingDirectory.isValid() shouldBe true` checks if directory is git repository, doesn't it?
Added check of revision.
</blockquote></details>

True.

---

*[downloader/src/funTest/kotlin/DownloaderTest.kt, line 150 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98#-L-akgI7vS93_egsJWDC:-L-bIkVTCExULAXm-bRA:betpdso) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/ecafbbc774b3044d1f1c39f6910ebf7df87d476d/downloader/src/funTest/kotlin/DownloaderTest.kt#L150)):*
<details><summary><i>Previously, pnowakowski wrote…</i></summary><blockquote>

I wasn't very happy  with how the tests are grouped, but didn't have a better idea. 
We already have MercurialTest, but there, Mercurial is being called directly, instead of with`Main.download(pkg, outputDir)` like it is here. 
I think we should either call Mercurial via `Main` class,  or call Git directly in here as well. 
Then we should add some tests for `Main.download(pkg, outputDir)` to cover cases that are not covered by Git and Mercurial tests.
I like refactoring MercurialTest option better. What do you think? 

</blockquote></details>

We could call tests that call Mercurial directly MercurialTest, and tests that call Main.download MercurialIntegrationTest. But now that I think about it I think we should call Mercurial directly in all Mercurial tests (same for Git), and have separate tests for Main.download which test its specific functionality. Most of it is in VersionControlSystem.forProvider which should also be tested separately (and already is partly).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98#-:-L-bKQw33u_2Fcil_BaX:b-mien9d)*

**Comment by mnonnenmacher** [comment]
Reviewed 1 of 2 files at r4, 4 of 4 files at r5.
Review status: 3 of 5 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/98)*
